File: 283. Move Zeroes.cpp
// Time:  O(n)
// Space: O(1)

// Solution 1
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int insertPos = 0;
        for (auto & num: nums) {
            if (num) {
                swap(nums[insertPos++], num);
            }
        }
    }
};

// Solution 2
class Solution {
public:
    void moveZeroes(vector<int>& nums) {
        int insertPos = 0;
        for (int i = 0; i < nums.size(); i++) {
            if (nums[i] != 0) {
                nums[insertPos++] = nums[i];
            }
        }
        while(insertPos < nums.size())
            nums[insertPos++] = 0;
    }
};

File: 283. Move Zeroes.py
# 283. Move Zeroes
# Time: O(N)
# Space: O(1)

# keep track of num zero index in the nums
class Solution(object):
    def moveZeroes(self, nums):
        """
        :type nums: List[int]
        :rtype: void Do not return anything, modify nums in-place instead.
        """
        idx = 0
        for num in nums:
            if num != 0:
                nums[idx] = num
                idx += 1
        for i in range(idx, len(nums)):
            nums[i] = 0
        